crates/ty_python_semantic/src/types/diagnostic.rs

```diff
@@ -3768,6 +3796,125 @@ pub(super) fn report_invalid_method_override<'db>(
+        "Cannot override final member `{member}` from superclass `{superclass_name}`"
+    ));
+
+    let mut sub = SubDiagnostic::new(
+        SubDiagnosticSeverity::Info,
+        format_args!(
+            "`{superclass_name}.{member}` is decorated with `@final`, forbidding overrides"
+        ),
+    );
+
+    let first_final_superclass_definition = superclass_method_defs
+        .iter()
+        .find(|function| function.has_known_decorator(db, FunctionDecorators::FINAL))
+        .expect(
+            "At least one function definition in the superclass should be decorated with `@final`",
+        );
+
+    let superclass_function_literal = if first_final_superclass_definition.file(db).is_stub(db) {
+        first_final_superclass_definition.first_overload_or_implementation(db)
+    } else {
+        first_final_superclass_definition
+            .literal(db)
+            .last_definition(db)
+    };
+
+    sub.annotate(
+        Annotation::secondary(Span::from(superclass_function_literal.focus_range(
+            db,
+            &parsed_module(db, first_final_superclass_definition.file(db)).load(db),
+        )))
+        .message(format_args!("`{superclass_name}.{member}` defined here")),
+    );
+
+    if let Some(decorator_span) =
+        superclass_function_literal.find_known_decorator_span(db, KnownFunction::Final)
+    {
+        sub.annotate(Annotation::secondary(decorator_span));
+    }
+
+    diagnostic.sub(sub);
+
+    let underlying_function = match subclass_type {
+        Type::FunctionLiteral(function) => Some(function),
+        Type::BoundMethod(method) => Some(method.function(db)),
+        _ => None,
+    };
+
+    if let Some(function) = underlying_function {
+        let overload_deletion = |overload: &OverloadLiteral<'db>| {
+            Edit::range_deletion(overload.node(db, context.file(), context.module()).range())
+        };
+
+        match function.overloads_and_implementation(db) {
+            ([first_overload, rest @ ..], None) => {
+                diagnostic.help(format_args!("Remove all overloads for `{member}`"));
+                diagnostic.set_fix(Fix::unsafe_edits(
+                    overload_deletion(first_overload),
+                    rest.iter().map(overload_deletion),
+                ));
+            }
+            ([first_overload, rest @ ..], Some(implementation)) => {
+                diagnostic.help(format_args!(
+                    "Remove all overloads and the implementation for `{member}`"
+                ));
+                diagnostic.set_fix(Fix::unsafe_edits(
+                    overload_deletion(first_overload),
+                    rest.iter().chain([&implementation]).map(overload_deletion),
+                ));
+            }
+            ([], Some(implementation)) => {
+                diagnostic.help(format_args!("Remove the override of `{member}`"));
+                diagnostic.set_fix(Fix::unsafe_edit(overload_deletion(&implementation)));
```

Comment:
A fix that might introduce invalid syntax should be a manual fix

Reply:
(fixed in https://github.com/astral-sh/ruff/pull/21699)